README.md
# Speck Cipher

[![Latest Version on Packagist](https://img.shields.io/packagist/v/deligoez/speck-cipher.svg?style=flat-square)](https://packagist.org/packages/deligoez/speck-cipher)
[![Tests](https://github.com/deligoez/speck-cipher/actions/workflows/run-tests.yml/badge.svg?branch=main)](https://github.com/deligoez/speck-cipher/actions/workflows/run-tests.yml)
[![Total Downloads](https://img.shields.io/packagist/dt/deligoez/speck-cipher.svg?style=flat-square)](https://packagist.org/packages/deligoez/speck-cipher)

This is where your description should go. Try and limit it to a paragraph or two. Consider adding a small example.

## Installation

You can install the package via composer:

```bash
composer require deligoez/speck-cipher
```

## Usage

```php
$skeleton = new Deligoez\Speck();
echo $skeleton->echoPhrase('Hello, Deligoez!');
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](https://github.com/spatie/.github/blob/main/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [Yunus Emre Deligöz](https://github.com/deligoez)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `speck-cipher` will be documented in this file.



tests/SpeckTest.php
<?php

use Deligoez\Speck\Speck;

test('Speck 32/64', function (): void {
    $key = [0x1918, 0x1110, 0x0908, 0x0100];
    $plainText = 0x6574694C;
    $cipherText = 0xA86842F2;

    $speck = new Speck(bits: 16, rounds: 22);

    $this->assertEquals($speck->encrypt($plainText, $key), $cipherText);
    $this->assertEquals($speck->decrypt($cipherText, $key), $plainText);
});

it('can test', function () {
    $speck = new Speck(
        bits: 16,
        rounds: 22,
        rightRotations: 7,
        leftRotations: 2,
    );

    $key = [0x0100, 0x0908, 0x1110, 0x1918];
    $originalInteger = 0x694C6574;
    $obfuscatedInteger = $speck->encrypt($originalInteger, $key);

    $this->assertEquals($obfuscatedInteger, 0x42F2A868);
    $this->assertEquals($originalInteger, $speck->decrypt($obfuscatedInteger, $key));
});

test('try', function (): void {
    $settings = [
        ['bits' => 8,  'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 10, 'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 15, 'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 16, 'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 20, 'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 24, 'rounds' => 22, 'rightRotations' => 8, 'leftRotations' => 3],
        ['bits' => 25, 'rounds' => 22, 'rightRotations' => 8, 'leftRotations' => 3],
        ['bits' => 26, 'rounds' => 22, 'rightRotations' => 8, 'leftRotations' => 3],
    ];

    foreach ($settings as $setting) {
        $keys = [];
        $maxKeyValue = 2 ** $setting['bits'];

        for ($i = 0; $i < 16; $i++) {
            $key = [];
            $keyLength = $i % 2 ? 4 : 2;

            for ($j = 0; $j < $keyLength; $j++) {
                $key[] = random_int(1, $maxKeyValue);
            }
            $keys[] = $key;
        }

        $testValues = [];
        $maxTestValue = 2 ** ($setting['bits'] * 2) - 1;
        for ($i = 0; $i < 256; $i++) {
            $testValues[] = random_int(1, $maxTestValue);
        }

        $speck = new Speck($setting['bits'], $setting['rounds'], $setting['rightRotations'], $setting['leftRotations']);
        foreach ($testValues as $testValue) {
            foreach ($keys as $key) {
                $cipher = $speck->encrypt($testValue, $key);
                $plain = $speck->decrypt($cipher, $key);

                $this->assertEquals($cipher, $plain);
            }
        }
    }
});


tests/SpeckV1Test.php
<?php

declare(strict_types=1);

use Deligoez\Speck\SpeckV1;

test('Speck 32/64 offical vector', function (): void {
    $keys = [0x1918, 0x1110, 0x0908, 0x0100];
    $plainText = [0x6574, 0x694C];
    $cipherText = [0xA868, 0x42F2];

    $speck = new SpeckV1(bits: 64, rounds: 22);

    dd($speck->encryptRaw($plainText, $keys));

//    $this->assertEquals($speck->encrypt($plainText, $key), $cipherText);
//    $this->assertEquals($speck->decrypt($cipherText, $key), $plainText);
});

test('Speck 32/64', function (): void {
    $key = [0x1918, 0x1110, 0x0908, 0x0100];
    $plainText = 0x6574694C;
    $cipherText = 0xA86842F2;

    $speck = new SpeckV1(bits: 16, rounds: 22);

    $this->assertEquals($speck->encrypt($plainText, $key), $cipherText);
    $this->assertEquals($speck->decrypt($cipherText, $key), $plainText);
});

it('can test', function () {
    $speck = new SpeckV1(
        bits: 16,
        rounds: 22,
        rightRotations: 7,
        leftRotations: 2,
    );

    $key = [0x0100, 0x0908, 0x1110, 0x1918];
    $originalInteger = 0x694C6574;
    $obfuscatedInteger = $speck->encrypt($originalInteger, $key);

    $this->assertEquals($obfuscatedInteger, 0x42F2A868);
    $this->assertEquals($originalInteger, $speck->decrypt($obfuscatedInteger, $key));
});

test('try', function (): void {
    $settings = [
        ['bits' => 8,  'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 10, 'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 15, 'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 16, 'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 20, 'rounds' => 22, 'rightRotations' => 7, 'leftRotations' => 2],
        ['bits' => 24, 'rounds' => 22, 'rightRotations' => 8, 'leftRotations' => 3],
        ['bits' => 25, 'rounds' => 22, 'rightRotations' => 8, 'leftRotations' => 3],
        ['bits' => 26, 'rounds' => 22, 'rightRotations' => 8, 'leftRotations' => 3],
    ];

    foreach ($settings as $setting) {
        $keys = [];
        $maxKeyValue = 2 ** $setting['bits'];

        for ($i = 0; $i < 16; $i++) {
            $key = [];
            $keyLength = $i % 2 ? 4 : 2;

            for ($j = 0; $j < $keyLength; $j++) {
                $key[] = random_int(1, $maxKeyValue);
            }
            $keys[] = $key;
        }

        $testValues = [];
        $maxTestValue = 2 ** ($setting['bits'] * 2) - 1;
        for ($i = 0; $i < 256; $i++) {
            $testValues[] = random_int(1, $maxTestValue);
        }

        $speck = new SpeckV1($setting['bits'], $setting['rounds'], $setting['rightRotations'], $setting['leftRotations']);
        foreach ($testValues as $testValue) {
            foreach ($keys as $key) {
                $cipher = $speck->encrypt($testValue, $key);
                $plain = $speck->decrypt($cipher, $key);

                $this->assertEquals($cipher, $plain);
            }
        }
    }
});


tests/SpeckValidationTest.php
<?php

declare(strict_types=1);

use Deligoez\Speck\Exceptions\InvalidBlockSizeException;
use Deligoez\Speck\Exceptions\InvalidKeySizeException;
use Deligoez\Speck\Speck;

it('throws InvalidBlockSizeException for invalid block sizes', function (): void {
    new Speck(key: 0x1918111009080100, blockSize: 1);
})->expectException(InvalidBlockSizeException::class);

it('throws InvalidKeySizeException for invalid key sizes', function (): void {
    new Speck(key: 0x1918111009080100, keySize: 1, blockSize: 128);
})->expectException(InvalidKeySizeException::class);


src/SpeckV1.php
<?php

declare(strict_types=1);

namespace Deligoez\Speck;

class SpeckV1
{
    protected int $bitMax;
    protected int $bitMask;

    public function __construct(
        protected int $bits = 16,
        protected int $rounds = 22,
        protected int $rightRotations = 7,
        protected int $leftRotations = 2,
    ) {
        $this->bitMax = 2 ** $this->bits;
        $this->bitMask = $this->bitMax - 1;
    }

    public function rotateRight(int $x, int $numberOfRotations): int
    {
        return ($x >> $numberOfRotations) | (($x << ($this->bits - $numberOfRotations)) & $this->bitMask);
    }

    public function rotateLeft(int $x, int $numberOfRotations): int
    {
        return (($x << $numberOfRotations) & $this->bitMask) | ($x >> ($this->bits - $numberOfRotations));
    }

    /**
     * Complete one round of Feistel operation.
     *
     * @return int[]
     */
    public function round(?int $x, int $y, int $k): array
    {
        $x = $this->rotateRight(x: $x, numberOfRotations: $this->rightRotations);
        $x = ($x + $y) & $this->bitMask;
        $x ^= $k;
        $y = $this->rotateLeft(x: $y, numberOfRotations: $this->leftRotations);
        $y ^= $x;

        return [$x, $y];
    }

    /**
     * Complete one round of inverse Feistel operation.
     */
    public function roundReverse(?int $x, int $y, int $k): array
    {
        $y ^= $x;
        $y = $this->rotateRight(x: $y, numberOfRotations: $this->leftRotations);
        $x ^= $k;
        $x = ($x - $y) & $this->bitMask;
        $x = $this->rotateLeft(x: $x, numberOfRotations: $this->rightRotations);

        return [$x, $y];
    }

    public function encryptRaw(array $plainTexts, array $keys): array
    {
        [$y, $x] = $plainTexts;
        $b = $keys[0];
        $a = array_slice($keys, offset: 1);

        [$x, $y] = $this->round(x: $x, y: $y, k: $b);

        $count = count($a);
        for ($i = 0; $i < $this->rounds - 1; $i++) {
            $j = $i % $count;
            [$a[$j], $b] = $this->round(x: $a[$j] ?? null, y: $b, k: $i);
            [$x, $y] = $this->round(x: $x, y: $y, k: $b);
        }

        return [$y, $x];
    }

    public function decryptRaw(array $cipherTexts, array $keys): array
    {
        [$y, $x] = $cipherTexts;
        $b = $keys[0];
        $a = array_slice($keys, offset: 1);

        for ($i = 0; $i < $this->rounds - 1; $i++) {
            $j = $i % count($a);
            [$a[$j], $b] = $this->round(x: $a[$j], y: $b, k: $i);
        }

        for ($i = 0; $i < $this->rounds; $i++) {
            $j = ($this->rounds - 2 - $i) % count($a);
            [$x, $y] = $this->roundReverse(x: $x, y: $y, k: $b);
            [$a[$j], $b] = $this->roundReverse(x: $a[$j] ?? null, y: $b, k: $this->rounds - 2 - $i);
        }

        return [$y, $x];
    }

    public function encrypt(int $input, array $keys): int
    {
        $result = $this->encryptRaw(
            plainTexts: [
                (int) ($input / $this->bitMax) | 0,
                $input & $this->bitMask,
            ],
            keys: $keys
        );

        return $result[0] * $this->bitMax + $result[1];
    }

    public function decrypt(int $input, array $keys): int
    {
        $result = $this->decryptRaw(
            cipherTexts: [
                (int) ($input / $this->bitMax) | 0,
                $input & $this->bitMask,
            ],
            keys: $keys
        );

        return $result[0] * $this->bitMax + $result[1];
    }
}


src/Speck.php
<?php

declare(strict_types=1);

namespace Deligoez\Speck;

use Deligoez\Speck\Exceptions\InvalidBlockSizeException;
use Deligoez\Speck\Exceptions\InvalidKeySizeException;
use GMP;

/**
 * Speck Block Cipher Object.
 */
class Speck
{
    // blockSize => [[keySize => numberOfRounds]]
    public const VALID_SETUPS = [
        32 => [64 => 22],
        48 => [72 => 22, 96 => 23],
        64 => [96 => 26, 128 => 27],
        96 => [96 => 28, 144 => 29],
        128 => [128 => 32, 192 => 33, 256 => 34],
    ];
    protected const VALID_MODES = ['ECB', 'CTR', 'CBC', 'PCBC', 'CFB', 'OFB'];

    /** Integer representation of the encryption key */
    protected int|GMP $key;

    /** Integer representing the size of the encryption key in bits */
    protected int $keySize;

    /** Integer representing the size of the blocks in bits */
    protected int $blockSize;

    /** Integer representing the size of the words in bits */
    protected int $wordSize;

    protected int $rounds;
    protected GMP $modMask;
    protected GMP $modMaskSub;

    /** Number of left rotations */
    protected int $betaShift;

    /** Number of right rotations */
    protected int $alphaShift;

    /** @var array<GMP> */
    protected array $keySchedule;

    /**
     * Initialize an instance of the Speck Block Cipher.
     *
     * @throws \Deligoez\Speck\Exceptions\InvalidBlockSizeException
     * @throws \Deligoez\Speck\Exceptions\InvalidKeySizeException
     */
    public function __construct(
        int|GMP $key,
        int $keySize = 128,
        int $blockSize = 128,
    ) {
        // Block size validation
        if (! array_key_exists($blockSize, self::VALID_SETUPS)) {
            InvalidBlockSizeException::build();
        }

        // Key size validation
        if (! array_key_exists($keySize, self::VALID_SETUPS[$blockSize])) {
            InvalidKeySizeException::build($blockSize);
        }

        $this->keySize = $keySize;
        $this->blockSize = $blockSize;

        $this->wordSize = $this->blockSize >> 1;
        $this->rounds = self::VALID_SETUPS[$this->blockSize][$keySize];

        // Mod mask for modular subtraction
        $this->modMaskSub = gmp_pow(num: 2, exponent: $this->wordSize);

        // Create properly sized bit mask for truncating addition and left shift outputs
        $this->modMask = gmp_sub(num1: $this->modMaskSub, num2: 1);

        // Setup circular shift parameters
        $this->betaShift = $this->blockSize === 32 ? 2 : 3;
        $this->alphaShift = $this->blockSize === 32 ? 7 : 8;

        // Parse the given key and truncate it to the key length
        $this->key = gmp_and($key, ((gmp_pow(2, $this->keySize)) - 1));

        // Pre-compile key schedule
        $this->keySchedule = [gmp_and($this->key, $this->modMask)];

        $lSchedule = [];
        $count = (int) ($this->keySize / $this->wordSize);
        for ($i = 1; $i < $count; $i++) {
            $lSchedule[] = gmp_and($this->gmp_shiftr($this->key, $i * $this->wordSize), $this->modMask);
        }

        for ($i = 0; $i < $this->rounds - 1; $i++) {
            [$lSchedule[], $this->keySchedule[]] = $this->round($lSchedule[$i], $this->keySchedule[$i], gmp_init($i));
        }
    }

    // region Encryption

    /**
     * Complete one round of Feistel operation.
     *
     * @param \GMP $upperWord Upper bits of the current plain text.
     * @param \GMP $lowerWord Lower bits of the current plain text.
     * @param \GMP $key       Round key.
     *
     * @return array{0: GMP, 1: GMP } Upper and lower cipher text segments.
     */
    protected function round(GMP $upperWord, GMP $lowerWord, GMP $key): array
    {
        $upperWord = gmp_and(gmp_add($this->gmp_shiftl($upperWord, ($this->wordSize - $this->alphaShift)), $this->gmp_shiftr($upperWord, $this->alphaShift)), $this->modMask);
        $upperWord = gmp_and(gmp_add($upperWord, $lowerWord), $this->modMask);
        $upperWord = gmp_xor($key, $upperWord);

        $lowerWord = gmp_and(gmp_add($this->gmp_shiftr($lowerWord, ($this->wordSize - $this->betaShift)), ($this->gmp_shiftl($lowerWord, $this->betaShift))), $this->modMask);
        $lowerWord = gmp_xor($upperWord, $lowerWord);

        return [$upperWord, $lowerWord];
    }

    /**
     * Completes appropriate number of Speck Feistel functions to encrypt provided words.
     * Round number is based off of number of elements in the key schedule.
     *
     * @param \GMP $upperWord Upper bits of the current plain text input. Limited by the word size of currenly configured cipher.
     * @param \GMP $lowerWord Lower bits of the current plain text input. Limited by the word size of currenly configured cipher.
     *
     * @return array{ 0: GMP, 1: GMP } Upper and lower bits of the cipher text words.
     */
    protected function encryptRaw(GMP $upperWord, GMP $lowerWord): array
    {
        foreach ($this->keySchedule as $key) {
            [$upperWord, $lowerWord] = $this->round($upperWord, $lowerWord, $key);
        }

        return [$upperWord, $lowerWord];
    }

    /**
     * Process a new plain text into the cipher text based on current cipher setup.
     *
     * @param \GMP|string|int $plainText Integer value to encrypt.
     *
     * @return \GMP Encrypted integer value
     */
    public function encrypt(GMP|string|int $plainText): GMP
    {
        $b = gmp_and($this->gmp_shiftr($plainText, $this->wordSize), $this->modMask);
        $a = gmp_and($plainText, $this->modMask);

        [$b, $a] = $this->encryptRaw($b, $a);

        return gmp_add($this->gmp_shiftl($b, $this->wordSize), $a);
    }

    // endregion

    // region Decryption

    /**
     * Complete one round of reverse Feistel operation.
     *
     * @param \GMP $upperWord Upper bits of the current cipher text.
     * @param \GMP $lowerWord Lower bits of the current cipher text.
     * @param \GMP $key       Round key.
     *
     * @return array{0: GMP, 1: GMP } Upper and lower plain text segments.
     */
    protected function reverseRound(GMP $upperWord, GMP $lowerWord, GMP $key): array
    {
        $upperWord = gmp_xor($upperWord, $lowerWord);
        $upperWord = gmp_and(gmp_add($this->gmp_shiftl($upperWord, ($this->wordSize - $this->betaShift)), ($this->gmp_shiftr($upperWord, $this->betaShift))), $this->modMask);
        $upperWord = gmp_xor($upperWord, $key);

        $lowerWord = gmp_mod(gmp_add(gmp_sub($upperWord, $lowerWord), $this->modMaskSub), $this->modMaskSub);
        $lowerWord = gmp_and(gmp_add($this->gmp_shiftr($lowerWord, ($this->wordSize - $this->alphaShift)), ($this->gmp_shiftl($lowerWord, $this->alphaShift))), $this->modMask);

        return [$upperWord, $lowerWord];
    }

    /**
     * Completes appropriate number of Speck Feistel functions to decrypt provided words.
     * Round number is based off of number of elements in the key schedule.
     *
     * @param \GMP $upperWord Upper bits of the current cipher text input. Limited by the word size of currenly configured cipher.
     * @param \GMP $lowerWord Lower bits of the current cipher text input. Limited by the word size of currenly configured cipher.
     *
     * @return array{ 0: GMP, 1: GMP } Upper and lower bits of the plain text words.
     */
    protected function decryptRaw(GMP $upperWord, GMP $lowerWord): array
    {
        foreach (array_reverse($this->keySchedule) as $key) {
            $this->reverseRound($upperWord, $lowerWord, $key);
        }

        return [$upperWord, $lowerWord];
    }

    /**
     * Process a cipher text into the plain text based on current cipher setup.
     *
     * @param \GMP|string|int $ciphertext Integer value to decrypt.
     *
     * @return \GMP Decrypted integer value
     */
    public function decrypt(GMP|string|int $ciphertext): GMP
    {
        $b = gmp_and($this->gmp_shiftr($ciphertext, $this->wordSize), $this->modMask);
        $a = gmp_and($ciphertext, $this->modMask);

        [$b, $a] = $this->decryptRaw($b, $a);

        return gmp_add($this->gmp_shiftl($b, $this->wordSize), $a);
    }

    // endregion

    // region Helpers

    /**
     * Bitwise left shift.
     * Moves all bits in the number to the left by the given number of places.
     *
     * @param \GMP|string|int $number         Number to bitwise left shift.
     * @param int             $numberOfShifts Number of places to shift.
     */
    protected function gmp_shiftl(GMP|string|int $number, int $numberOfShifts): GMP
    {
        return gmp_mul($number, gmp_pow(2, $numberOfShifts));
    }

    /**
     * Bitwise right shift.
     * Moves all bits in the number to the right by the given number of places.
     *
     * @param \GMP|string|int $number         Number to bitwise right shift.
     * @param int             $numberOfShifts Number of places to shift.
     */
    protected function gmp_shiftr(GMP|string|int $number, int $numberOfShifts): GMP
    {
        return gmp_div($number, gmp_pow(2, $numberOfShifts));
    }

    // endregion
}
